lib/web/headless.go

```diff
@@ -0,0 +1,96 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+
+*/
+
+package web
+
+import (
+	"net/http"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+)
+
+func (h *Handler) getHeadless(_ http.ResponseWriter, r *http.Request, params httprouter.Params, sctx *SessionContext) (any, error) {
+	headlessAuthenticationID := params.ByName("headless_authentication_id")
+	if headlessAuthenticationID == "" {
+		return nil, trace.NotFound("failed to find Headless session") // TODO this or just failed?
+	}
+
+	authClient, err := sctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	headlessAuthn, err := authClient.GetHeadlessAuthentication(r.Context(), headlessAuthenticationID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return headlessAuthn, nil
+}
+
+func (h *Handler) putHeadlessState(w http.ResponseWriter, r *http.Request, params httprouter.Params, sctx *SessionContext) (any, error) {
+	headlessAuthenticationID := params.ByName("headless_authentication_id")
+	if headlessAuthenticationID == "" {
+		return nil, trace.NotFound("failed to find Headless session") // TODO this or just failed?
+	}
+
+	var req client.HeadlessRequest
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var action types.HeadlessAuthenticationState
+	var resp = &proto.MFAAuthenticateResponse{}
+
+	switch req.Action {
+	case "accept":
+		action = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED
+		resp = &proto.MFAAuthenticateResponse{
+			Response: &proto.MFAAuthenticateResponse_Webauthn{
+				Webauthn: wanlib.CredentialAssertionResponseToProto(req.WebauthnAssertionResponse),
+			},
+		}
+	case "denied":
+		action = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_DENIED
+	default:
+		return nil, trace.BadParameter("unknown action %s", req.Action)
+	}
+
+	authClient, err := sctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = authClient.UpdateHeadlessAuthenticationState(r.Context(), headlessAuthenticationID,
+		action, resp)
+	if err != nil {
+		return nil, trace.Wrap(err) // TODO replace with failed to authenticate always?
+	}
+
+	// TODO(jakule): webui expects JSON on POST ¯\_(ツ)_/¯
+	w.Write([]byte("{\"status\": \"OK\"}"))
```

Comment:
The fetch utility that we use automatically parses the request as JSON, I've been wanting to change this but that's why you're needing this.

Reply:
@hatched Should we just leave it as it is, as UI cannot handle it without returning any response, or should the UI use a different helper function? For me sounds a little bit pointless to return something just because our UI crashes.

Reply:
We should definitely change that requirement in the UI as I've also frequently run up against the same issue but I feel like that's a separate task to this and may have unintended consequences.